microbench/src/main/java/alluxio/grpc/ZeroCopyReadResponseBench.java

```diff
@@ -0,0 +1,170 @@
+ * with varying sizes of chunk data and compares against a default implementation
+ * which does not implement zero-copy.
+ */
+@Fork(value = 1)
+@Warmup(iterations = 2, time = 3)
+@Measurement(iterations = 5, time = 3)
+@BenchmarkMode(Mode.Throughput)
+public class ZeroCopyReadResponseBench {
+
+  // Dumb OutputStream that consumes serialized bytes
+  private static final OutputStream SINK = new NullOutputStream();
+
+  // Custom marshaller
+  private static final ReadResponseMarshaller ZERO_COPY_MARSHALLER = new ReadResponseMarshaller();
+
+  // Default marshaller that comes with the protobuf definition
+  private static final MethodDescriptor.Marshaller<ReadResponse> DEFAULT_MARSHALLER =
+          ProtoUtils.marshaller(ReadResponse.getDefaultInstance());
+
+  // Buffer used to drain InputStream manually
+  private static final byte[] BUF = new byte[4096];
+
+  @State(Scope.Benchmark)
+  public static class BenchParams {
+    // marshal a read response with big chunk size
+    // 1MB, 10MB, 20MB, 50MB, 100MB
+    @Param({ "1", "10", "20", "50", "100" })
+    public int mChunkSizeMB;
+
+    // random byte generator
+    private final Random mRandom = new Random();
+
+    public int mChunkSizeByte;
+
+    // chunk data buffer
+    public DataBuffer mChunkData;
+
+    // ReadResponse object used for marshalling
+    public ReadResponse mReadResponse;
+
+    // Serialized InputStream used for unmarshalling
+    public InputStream mReadResponseInputStream;
+
+    @Setup(Level.Invocation)
+    public void setup() {
+      mChunkSizeByte = mChunkSizeMB * 1024 * 1024;
+
+      // set up chunk of data
+      byte[] bytes = new byte[mChunkSizeByte];
+      mRandom.nextBytes(bytes);
+      mChunkData = new NettyDataBuffer(Unpooled.wrappedBuffer(bytes));
+
+      // set up response object
+      mReadResponse = ReadResponse
+              .newBuilder()
+              .setChunk(Chunk.newBuilder().setData(UnsafeByteOperations.unsafeWrap(bytes)))
+              .build();
+
+      // set up serialized stream
+      mReadResponseInputStream = new ByteArrayInputStream(mReadResponse.toByteArray());
+    }
+  }
+
+  @Benchmark
+  public void marshalZeroCopy(BenchParams params) throws IOException {
+    ZERO_COPY_MARSHALLER.offerBuffer(params.mChunkData, params.mReadResponse);
+    try (InputStream is = ZERO_COPY_MARSHALLER.stream(params.mReadResponse)) {
+      ((Drainable) is).drainTo(SINK);
+    }
+  }
+
+  @Benchmark
+  public void marshalBaselineDrain(BenchParams params) throws IOException {
+    try (InputStream is = DEFAULT_MARSHALLER.stream(params.mReadResponse)) {
+      ((Drainable) is).drainTo(SINK);
+    }
+  }
+
+  @Benchmark
+  public void marshalBaselineRead(BenchParams params) throws IOException {
+    try (InputStream is = DEFAULT_MARSHALLER.stream(params.mReadResponse)) {
+      int byteRead;
+      while ((byteRead = is.read(BUF)) != -1) {
+        SINK.write(BUF, 0, byteRead);
+      }
+    }
+  }
```

Comment:
The custom marshaller returns a `ProtoInputStream` when called `stream`, which can both act as an `InputStream` and as a `Drainable`. Both ways of interaction are benchmarked.